### PR TITLE
Add API key RPCs

### PR DESCRIPTION
**Pull request by bduffany**
* Add and implement the RPCs for all the request/response protos previously added
* Remove all references to `group.APIKey` in favor of the `APIKeys` table (e.g. in `ServeHTTP`), except for the refs in `GetBazelConfig` (which are handled in another PR)
* Add a DB migration to insert an `APIKeys` row for all existing `Group` rows

### CODE REVIEW

File: server/buildbuddy_server/buildbuddy_server.go [hunk 1]
```diff
@@ -381,6 +382,114 @@ func (s *BuildBuddyServer) JoinGroup(ctx context.Context, req *grpb.JoinGroupReq
...
+	return &akpb.CreateApiKeyResponse{
+		ApiKey: &akpb.ApiKey{
+			Id:    k.APIKeyID,
+			Value: k.Value,
+			Label: k.Label,
+		},
+	}, nil
+}
+
+func (s *BuildBuddyServer) authorizeAPIKeyAccess(ctx context.Context, apiKeyID string) error {
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:433]: can you replace this with a standard permissions check?

File: server/tables/tables.go [hunk 2]
```diff
@@ -243,6 +243,24 @@ func (t *Token) TableName() string {
 	return "Tokens"
 }
 
+type APIKey struct {
+	Model
+
+	// Identifiers
+
```
Comment (tylerwilliams) [inline server/tables/tables.go:250]: nit: can you rm the newlines after // comments?

File: server/buildbuddy_server/buildbuddy_server.go [hunk 3]
```diff
@@ -381,6 +382,114 @@ func (s *BuildBuddyServer) JoinGroup(ctx context.Context, req *grpb.JoinGroupReq
...
+	}
+	groupID := req.GetGroupId()
+	if err := s.authorizeGroupAccess(ctx, groupID); err != nil {
+		return nil, err
+	}
+	tableKeys, err := userDB.GetAPIKeys(ctx, groupID)
+	if err != nil {
+		return nil, err
+	}
+	protoKeys := make([]*akpb.ApiKey, len(tableKeys))
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:398]: nice -- did you read about the difference between length and capacity?

In this code what you have is just perfect, but if you occasionally expected an error in your for loop that would mean skipping an element, you could instead do:

keys := make([].., 0, len(tableKeys))
for {
  keys = append(keys, blah)
}

and still get most of the benefit of knowing the length ahead of time (aka reduced allocations).

**Review by tylerwilliams — CHANGES_REQUESTED** [review]
Talked offline -- I think we can handle the perms in a simpler-to-maintain way.

Overall this CL looks really good!

File: server/buildbuddy_server/buildbuddy_server.go (hunk 1 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go]: Done

File: server/buildbuddy_server/buildbuddy_server.go (hunk 3 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go:395]: Ah cool, I knew there was a difference, somehow missed that you could specify the capacity explicitly in `make` -- done!

File: server/buildbuddy_server/buildbuddy_server.go [hunk 4]
```diff
@@ -381,6 +379,106 @@ func (s *BuildBuddyServer) JoinGroup(ctx context.Context, req *grpb.JoinGroupReq
...
+	for _, k := range tableKeys {
+		protoKeys = append(protoKeys, &akpb.ApiKey{
+			Id:    k.APIKeyID,
+			Value: k.Value,
+			Label: k.Label,
+		})
+	}
+	return &akpb.GetApiKeysResponse{
+		ApiKey: protoKeys,
+	}, nil
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:405]: This is fine like you have it, but you can also just append these directly to your response like:

rsp := &akpb.GetApiKeysResponse{
	ApiKey: make([]*akpb.ApiKey, 0, len(tableKeys)),
}
for ... {
  rsp.ApiKey = append(rsp.ApiKey, key{})
}

return rsp, nil

File: server/buildbuddy_server/buildbuddy_server.go [hunk 5]
```diff
@@ -381,6 +379,106 @@ func (s *BuildBuddyServer) JoinGroup(ctx context.Context, req *grpb.JoinGroupReq
...
+	}
+	if err := s.authorizeAPIKeyWrite(ctx, req.GetId()); err != nil {
+		return nil, err
+	}
+	if err := userDB.UpdateAPIKey(ctx, &tables.APIKey{
+		APIKeyID: req.GetId(),
+		Label:    req.GetLabel(),
+	}); err != nil {
+		return nil, err
+	}
```
Comment (tylerwilliams) [inline server/buildbuddy_server/buildbuddy_server.go:464]: super uber nit: for readability's sake I think it's nice to keep the if cases to one line as much as possible. In this case you could do:

tk := &tables.APIKey{
	APIKeyID: req.GetId(),
	Label:    req.GetLabel(),
}
if err := userDB.UpdateAPIKey(ctx, tk); err != nil {
  return nil, err
}

File: server/tables/tables.go [hunk 6]
```diff
@@ -243,6 +243,25 @@ func (t *Token) TableName() string {
 	return "Tokens"
 }
 
+type APIKey struct {
+	Model
+
+	APIKeyID     string `gorm:"primary_key"`
+	GroupGroupID string `gorm:"index:group_id_index"`
```
Comment (tylerwilliams) [inline server/tables/tables.go:250]: indexes in gorm are (annoyingly) in the global namespace by default. So I think we should prefer to prefix index names with the table names to keep things sane.

So can you change group_id_index and value_index to...

api_key_group_id_index
api_key_value_index

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: server/buildbuddy_server/buildbuddy_server.go (hunk 4 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go]: Done

File: server/buildbuddy_server/buildbuddy_server.go (hunk 5 above)
Comment (bduffany) [reply server/buildbuddy_server/buildbuddy_server.go]: Much cleaner - done

File: server/tables/tables.go (hunk 6 above)
Comment (bduffany) [reply server/tables/tables.go]: Ah, unfortunate. Done

**Review by tylerwilliams — APPROVED** [review]
LGTM

File: server/tables/tables.go [hunk 7]
```diff
@@ -243,6 +243,24 @@ func (t *Token) TableName() string {
...
 
+type APIKey struct {
+	Model
+
+	APIKeyID string `gorm:"primary_key"`
+	UserID   string
+	GroupID  string `gorm:"index:api_key_group_id_index"`
+	Perms    int
+	// The API key token used for authentication.
+	Value string `gorm:"unique_index:api_key_value_index"`
```
Comment (tylerwilliams) [inline server/tables/tables.go:254]: Is there a reason to separate the Value from the APIKeyID since they are both strings and both the same value?

Comment (bduffany) [reply server/tables/tables.go:255]: Will double check the code that I wrote but I think `APIKeyID` and `Value` should be different - the ID follows the standard `PrimaryKeyForTable` convention (`AK12345...`) and the `Value` is the 20-char alphanumeric token that we generate, and the one that's actually used for auth

**Pull request merged**